### ISSUE TITLE
Tracepoints Don't Work

### ISSUE THREAD

**Issue by probably-not** [runsidekick/sidekick-agent-nodejs#2]
I installed sidekick on my Node.js application, and set up the VSCode extension (not using the cloud IDE, full read access to all public and private repositories that I own without any way of restricting the repositories that you have access to is a hard no).

Following the docs, I added a LogPoint, which worked fine, but when I added a TracePoint, I see that it never arrives in the Trace Events table. I turned on Debug logs, and whenever the tracepoint tries to send an event, it fails with the following error:
```
debug <TracePointAction> An error occured while resolving frames Converting circular structure to JSON
debug     --> starting at object with constructor 'Socket'
debug     |     property 'parser' -> object with constructor 'HTTPParser'
debug     --- property 'socket' closes the circle
debug <CommunicationManager> Sending data to communication api.
debug <TracePointAction> Tracepoint probe action working ...
debug <DefaultDebugApi> Script information can not be empty.
```

This fails consistently on all TracePoints I try to add.

**Comment by probably-not** [comment]
My guess is that this is coming from this line: https://github.com/runsidekick/sidekick-agent-nodejs/blob/master/packages/sidekick-agent-nodejs/src/api/internal/debug/converter/CaptureFrameConverter.ts#L33
Which is converting all of the `locals` in the frame into JSON using `JSON.stringify`.

**Comment by probably-not** [comment]
The official MDN docs recommend either utilizing a library that deals with circular references or using the example that they provide on their docs page to drop repeated references completely: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Cyclic_object_value#circular_references

**Comment by boroskoyo** [comment]
Hi there, thank you for this issue. We are on it and will fix it in a short time. We have spotted that problematic code is introduced with the 0.0.9 version. Until we fix this issue you can use 0.0.8 which lacks data reaction and open telemetry but works fine with tracepoints.

**Comment by serkan-ozal** [comment]
Hi @probably-not, Could you please try with version `0.0.13` and let us know whether it fixes your issue?

**Comment by bcaglayan** [comment]
Seems to be fixed on version 0.0.13.

**Comment by probably-not** [comment]
@serkan-ozal @bcaglayan While it looks like the cyclic JSON issue was solved, I'm still not receiving any tracepoint data in VSCode.

I see that whenever I pass the tracepoint, the debug log shows this:
```
debug <TracePointAction> Tracepoint probe action working ...
debug <DefaultDebugApi> Script information can not be empty.
debug <CommunicationManager> Sending data to communication api.
```

I'm also now no longer receiving LogPoints that worked previously, with similar logs to the tracepoint logs:
```
debug <LogPointAction> Logpoint probe action working ...
debug <DefaultDebugApi> Script information can not be empty.
```

**Comment by bcaglayan** [comment]
Hello @probably-not, i will try to reproduce issue

- what is your node.js version ? 
- do you use typescript ?
- what is your os ?
- can you share your app (your test app or project folder structure) ?

**Comment by probably-not** [comment]
@bcaglayan  My mistake, I believe it is working now, I've figured out my mistake.

In the VSCode extension the `Sidekick: Trace Events` window is always open and available (same with the `Sidekick: Log Events` window). This leads to the UX and expectation that if I have a tracepoint/logpoint in my code, I'll be able to see it in the window.

However, the window stays empty unless you explicitly click on a specific tracepoint/logpoint, at which point it does populate with all of the events that occurred up until that point.

Is there a repository to open an issue for the VSCode extension to fix this behavior? It's very confusing and not explained at all in the VSCode Extension UI anywhere, so you get stuck not knowing if there's an issue or not until you know that you have to click.

**Comment by probably-not** [comment]
While the TracePoint receiving is now working, I do see some more issues. My understanding is that a TracePoint is supposed to show all of the local variables that have been instantiated up to that point, similar to a normal debugger, is that correct? If yes, there seems to be an issue there, as the only values I'm seeing in the TracePoint is the function arguments and not the actual local variables, which kind of makes the feature useless since I can't debug...

To answer your questions about the setup:
- NodeJS version 16 (minor version is whatever the latest is on `gcr.io/distroless/nodejs:16` Docker image)
- We do use TypeScript
- OS is running in Docker on `gcr.io/distroless/nodejs:16`
- My current setup is a basic [NestJS](https://nestjs.com/) project, and I'm adding Tracepoints and Logpoints in a [Catch All Exception Filter](https://docs.nestjs.com/exception-filters) and simply triggering it via an endpoint in my API. I can't share the actual repo (proprietary) but I will try to create an example repo to reproduce the missing local variables issue
- The logs are printing out exactly what I have already sent here: https://github.com/runsidekick/sidekick-agent-nodejs/issues/2#issuecomment-1233809667

**Comment by serkan-ozal** [comment]
> While the TracePoint receiving is now working, I do see some more issues. My understanding is that a TracePoint is supposed to show all of the local variables that have been instantiated up to that point, similar to a normal debugger, is that correct? If yes, there seems to be an issue there, as the only values I'm seeing in the TracePoint is the function arguments and not the actual local variables, which kind of makes the feature useless since I can't debug...

Hi @probably-not , do you mean seeing local variables of all frames up to bottom? Do you ss local variables of the top frame (method)?

**Comment by bcaglayan** [comment]
@probably-not i created an issue about VSCode plugin tabs auto refresh themselves issues. You can follow-up state on [this issue link](https://github.com/runsidekick/sidekick-agent-nodejs/issues/4)

**Comment by probably-not** [comment]
@serkan-ozal The only values I'm seeing in the tracepoint data are the function argument values. For example, I have a function that looks something like [the catch all example on NestJS's docs site](https://docs.nestjs.com/exception-filters#catch-everything):
```typescript
catch(exception: unknown, host: ArgumentsHost): void {
    const ctx = host.switchToHttp();

    const httpException =
      exception instanceof HttpException
        ? (exception as HttpException)
        : undefined;
    ...
  }
```

I put the tracepoint on the `const httpException` line, but the event only contains two keys, `exception` and `host` (the function arguments). I don't see `ctx`, which is a local that has been initialized in the function already.

**Comment by bcaglayan** [comment]
> > While the TracePoint receiving is now working, I do see some more issues. My understanding is that a TracePoint is supposed to show all of the local variables that have been instantiated up to that point, similar to a normal debugger, is that correct? If yes, there seems to be an issue there, as the only values I'm seeing in the TracePoint is the function arguments and not the actual local variables, which kind of makes the feature useless since I can't debug...
> 
> Hi @probably-not , do you mean seeing local variables of all frames up to bottom? Do you ss local variables of the top frame (method)?

If do you want to see local variables of caller stacks (all frames), you can set `SIDEKICK_AGENT_MAX_EXPAND_FRAMES` environment variable. Its default value is 1.

_Note: Every opened frame will bring extra load._

**Comment by serkan-ozal** [comment]
Hi @probably-not, 

Had you a chance to put tracepoint to the next line to see whether it help to get `ctx` variable snapshot.

Also is it possible to share the project or any code that can reproduce the issue?

**Comment by probably-not** [comment]
Hi @serkan-ozal, I was able to test this morning by putting the tracepoint on a the line after the `const httpException` initialization. What I see now is 2 empty frames, and one frame that contains `exception`, `host`, `_a` (which I don't know where it is coming from), and `httpException`. It is still missing the `ctx` variable, and I have no clue where it's getting `_a` and the empty frames.

Screenshot from the tracepoint itself:
![image](https://user-images.githubusercontent.com/41164052/188301734-12adf739-d4a3-4f10-964a-f3afde70197a.png)

**Comment by serkan-ozal** [comment]
Hi @probably-not,

What do you see in `_a`?

**Comment by probably-not** [comment]
@serkan-ozal

The current code looks like this:
```typescript
  catch(exception: unknown, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const httpAdapter = this.httpAdapterHost?.httpAdapter;

    const httpException =
      exception instanceof HttpException
        ? (exception as HttpException)
        : undefined;

...
}
```

Tracepoint is on the `const httpException = ` line.

This is what I'm seeing in the Tracepoint:
<img width="621" alt="image" src="https://user-images.githubusercontent.com/41164052/188428391-8adfe7d4-0470-47a7-aacb-9607ecb4b7f0.png">


It's missing the `ctx` variable, and it looks like maybe the `_a` value in contains the `httpAdapter` variable.

**Comment by bcaglayan** [comment]
When you debug the code (on ide), which variables do you see ? I created a basic [setup like yours](https://docs.nestjs.com/exception-filters#catch-everything) and i reproduced missing variable issue (ctx) and i 'm working o it. But i didn't see '_a' variable on sidekick.

**Comment by probably-not** [comment]
Debugging through node's debugger at the same breakpoint gives me this:
<img width="445" alt="image" src="https://user-images.githubusercontent.com/41164052/188441697-b0e4b6fe-bacb-4cba-bfde-5ffa581d2561.png">

I do see `ctx` and `httpAdapter` (neither were shown in sidekick's tracepoint). Interestingly, it looks like `_a` is in here as well. I guess Node is expanding the optional `?` parameter into `_a` for the check and it's coming in as a local in the context?

**Comment by probably-not** [comment]
I should also note, node's debugger gives a lot more information, including the class type of each variable, which sidekick currently just parses as `object` for everything.

**Comment by bcaglayan** [comment]
> Debugging through node's debugger at the same breakpoint gives me this: <img alt="image" width="445" src="https://user-images.githubusercontent.com/41164052/188441697-b0e4b6fe-bacb-4cba-bfde-5ffa581d2561.png">
> 
> I do see `ctx` and `httpAdapter` (neither were shown in sidekick's tracepoint). Interestingly, it looks like `_a` is in here as well. I guess Node is expanding the optional `?` parameter into `_a` for the check and it's coming in as a local in the context?

It probably seems to be related to the "@" operator as you said. About 'ctx' issue, i'm working on it. For obtain class type, i can't sure about we can obtain type of class on our snapshot process. (not exactly the same as the debugging process, but i will check are there any way to obtain type)

**Comment by serkan-ozal** [comment]
Hi @probably-not,

We have released version `0.0.14` with the fixes of following issues:
- JSON circular structure
- missing variable
- ambiguous object type

Could you please retry with that version and let us know whether it fixes your problem?
Once you give OK, we can close the issue.

**Comment by probably-not** [comment]
Hi @serkan-ozal , that's great news! I'm currently OOO until next Tuesday, so I'll do the updates and check it out to ensure everything is working then

**Comment by probably-not** [comment]
Hey @serkan-ozal , I was just able to check the new version. It looks good in terms of the missing variables ( see all of them now), but I am still getting 2 empty frames on the tracepoint:
![image](https://user-images.githubusercontent.com/41164052/189839018-62d059f3-c20d-4de3-9d30-a52f7c3b0457.png)

I'm not sure if this is a related issue or if I should open a new one

**Comment by serkan-ozal** [comment]
Hi @probably-not , 

In fact it is not a problem but default behaviour. By default we collect variables only at the top frame. To collect variables from other frames (lower frames), you need to set `SIDEKICK_AGENT_MAX_EXPAND_FRAMES` env var (default value is `1`) to higher value.

**Comment by probably-not** [comment]
Hi @serkan-ozal,

I understand that in order to see the variables in other frames I would have to expand more frames, but it seems like a bug that you show empty frames in the view. It makes it look as though there's things that came through and were somehow filtered out (not great for UX).

If you'd like, I can open a separate issue for removing empty frames from the view, since it's no longer related to this specific tracepoints issue.

**Comment by serkan-ozal** [comment]
Hi @probably-not, if you are able to see variables at the lower frames when you set `SIDEKICK_AGENT_MAX_EXPAND_FRAMES` env var, I think we are good to close this issue and you can create another issue for empty frames

**Comment by probably-not** [comment]
Alright, in that case, for now I will close this and create a separate issue for the frames. Thanks for the work on this!
